src/nvim/tui/tui.c

```diff
@@ -1440,13 +1440,28 @@ static void patch_terminfo_bugs(TUIData *data, const char *term,
   // terminfo, is a fixup, not an augmentation.
   data->unibi_ext.reset_cursor_style = unibi_find_ext_str(ut, "Se");
   data->unibi_ext.set_cursor_style = unibi_find_ext_str(ut, "Ss");
+
+  // Some terminals can not currently be trusted to report if they support
+  // DECSCUSR or not. So we need to have a blacklist for when we should not
+  // trust the reported features.
+  if( (vte_version != 0 && vte_version < 3900) || konsole ) {
+    data->unibi_ext.reset_cursor_style = -1;
+    data->unibi_ext.set_cursor_style = -1;
+  }
   if (-1 == data->unibi_ext.set_cursor_style) {
```

Comment:
That would explain why the existing code was checking `if (-1 == data->unibi_ext.set_cursor_style)`.